### PR TITLE
Fixed _upcast_err docstring and comments in _axes.py

### LINKED ISSUE #21901: Suggested clarification of comments in errorbar helpers
@efiring suggested a number of improvements to the errorbar xerr/yerr processing in https://github.com/matplotlib/matplotlib/pull/21884 that were deferred in interest of getting 3.5.1 out the door.

### PR DESCRIPTION
**Pull request by andrew-fennell**
## PR Summary
Addresses @efiring's concerns about the docstring and comments in #21884 (issue #21901).

Closes #21901

## PR Checklist

**Tests and Styling**
- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [N/A] New features are documented, with examples if plot related.
- [N/A] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [N/A] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [N/A] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

### CODE REVIEW

File: lib/matplotlib/axes/_axes.py [hunk 1]
```diff
@@ -3363,14 +3354,17 @@ def _upcast_err(err):
...
+                # Get the type of the first element
+                # Promote the outer container to match the inner container
                 atype = type(cbook.safe_first_element(err))
-                # you can not directly pass data to the init of `np.ndarray`
                 if atype is np.ndarray:
+                    # Converts using np.asarray, because data cannot
+                    # be directly passed to init of `np.ndarray`
                     return np.asarray(err, dtype=object)
-                # but you can for unyt and astropy uints
+                # Directly pass data to the init of np.ndarray, for cases
```
Comment (tacaswell) [inline lib/matplotlib/axes/_axes.py:3364]: We are only here if  `atype` is _not_ `np.ndarray`

File: lib/matplotlib/axes/_axes.py [hunk 2]
```diff
@@ -3363,14 +3354,17 @@ def _upcast_err(err):
                     # fails.
                     isinstance(cbook.safe_first_element(err), np.ndarray)
             ):
-                # grab the type of the first element, we will try to promote
-                # the outer container to match the inner container
+                # Get the type of the first element
+                # Promote the outer container to match the inner container
```
Comment (tacaswell) [inline lib/matplotlib/axes/_axes.py:3358]: The actual promotion happens in the next code block.

**Comment by andrew-fennell** [comment]
@tacaswell I pushed updates to address your comments. 

I saw #21901 was tagged as 3.5.2 and had been dormant for a while. I only realized that it was assigned to you after I had already made my changes. I hope it is okay that I worked on this.

File: lib/matplotlib/axes/_axes.py [hunk 3]
```diff
@@ -3363,14 +3354,18 @@ def _upcast_err(err):
                     # fails.
                     isinstance(cbook.safe_first_element(err), np.ndarray)
             ):
-                # grab the type of the first element, we will try to promote
-                # the outer container to match the inner container
+                # Get the type of the first element
                 atype = type(cbook.safe_first_element(err))
-                # you can not directly pass data to the init of `np.ndarray`
+                # If atype is np.ndarray, promote the outer container to
```
Comment (tacaswell) [inline lib/matplotlib/axes/_axes.py:3359]: In all cases we want to promote the outer to match the type of the inner.  The problem we are trying to solve is to turn `[TYPE([1, 2]), TYPE([3, 4]]` -> `TYPE([[1, 2], [3, 4]])`  where TYPE may be a ndarray subclass or something more exotic.  

If we were to just do `np.asarray(...)` we would drop any type information that the containers were carrying,  just get naive floats, and lose the information about what TYPE was or any extra metadata it was carrying.

If we do `TYPE([...])` then it fails for numpy due to ndarray not having a `__init__` signature that works with that!

**Review by tacaswell — APPROVED** [review]
Thank you for tolerating my comments on this.

**Comment by andrew-fennell** [comment]
> Thank you for tolerating my comments on this.

I think it's better to correct it right now instead of needing to fix it later!

**Comment by QuLogic** [comment]
meeseeksdev backport to v3.5.x

**Pull request merged**